_posts/2021-09-30-spring-i18n.md

```diff
@@ -0,0 +1,379 @@
+
+An object of the Java Locale class consists of at least a language code and a country or region code. used for identifying objects, not a container for the objects themselves. 
+
+The language codes are lower-case, two-letter codes defined by ISO-639. 
+
+The country codes are upper-case, two-letter codes defined by ISO-3166. 
+
+The Java Locale class also contains convenient constants for commonly used locales. For example, the `Locale.US` creates a Locale object for the United States.
+
+## Creating the Spring Boot Application for Internationalization
+
+To work with some examples of internationalization, let us first create a Spring Boot project with the help of the [Spring boot Initializr](https://start.spring.io/#!type=maven-project&language=java&platformVersion=2.5.4&packaging=jar&jvmVersion=11&groupId=io.pratik.i18n&artifactId=productapp&name=productapp&description=Demo%20project%20for%20Spring%20Boot%20Internationalization&packageName=io.pratik.i18n.productapp), and then open the project in our favorite IDE. We have not added any dependencies to Maven `pom.xml` since the scheduler is part of the core module of the Spring framework.
+
+We will create a web application with this project using [Spring Web MVC](https://docs.spring.io/spring-framework/docs/3.2.x/spring-framework-reference/html/mvc.html) framework which will render an HTML page in different languages depending on the user's language selection.
+
+## Steps for Internationalization
+
+Internationalization of applications broadly follows the below steps:
+
+1. Resolving the user's preferred locale from the incoming request from the user either in the form of a request parameter, cookies, or a request header.
+2. Intercepting the change of locale in the incoming request and storing it in the user's session or cookies.
+3. Defining locale-specific resources, for example, language files for supported languages.
+4. Map region and language-sensitive elements in the view to elements capable of reading content at runtime based on the user's language and region.  
+
+Let us look at these steps in detail in the following sections:
+
+## Resolving the Locale with LocaleResolver
+We use the `LocaleResolver` interface for resolving the locale of a user from the incoming request. 
+
+Spring provides the following implementations of the `LocaleResolver` interface that determine the current locale based on the session, cookies, the `Accept-Language` header, or sets the locale to a fixed value:
+
+- **FixedLocaleResolver**: Mostly used for debugging purposes. It resolves the locale to a fixed language mentioned in the `application. properties`.
+- **AcceptHeaderLocaleResolver**: Resolves the locale using an `accept-language` HTTP header retrieved from an HTTP request. 
+- **SessionLocaleResolver**: Resolves the locale and stores it in the HttpSession of the user. 
```

Comment:
Where does it resolve the locale from, though? Same question for the `CookieLocaleResolver`.